things you could do with mix - part 3 (#1)
drumusician/clock: blog/part_3_option_parser.md

## Things you could do with mix

### Part 3: OptionParser

> Hello again! This is part 3 of my blog post exploring things you could do with mix. If you haven't read the first two you can find those and [here](https://www.theguild.nl/things-you-could-do-with-mix/) and [here](https://www.theguild.nl/thing-you-could-do-with-mix-2/). Last time we explored some new DateTime features in Elixir 1.8.0 and changed our output based on when the new year had arrived. The test we wrote for that was very brittle as it relied on the actual time. So let's start by making sure our full test suite is green again.

Reviewer: ```suggestion
Hello again! This is part 3 of my blog post series exploring things you could do with mix. If you haven't read the first two you can find those [here](https://www.theguild.nl/things-you-could-do-with-mix/) and [here](https://www.theguild.nl/thing-you-could-do-with-mix-2/). Last time we explored some new DateTime features in Elixir 1.8.0 and changed our output based on when the new year had arrived. The test we wrote for that was very brittle as it relied on the actual time. So let's start by making sure our full test suite is green again.
```

  day_of_year =
    current_date_time_in_time_zone
    |> DateTime.to_date()
    |> Date.day_of_year()

  city =
    String.split(time_zone, "/")
    |> Enum.drop(1)
    |> to_string()

  time =
    current_date_time_in_time_zone
    |> DateTime.truncate(:second)
    |> DateTime.to_time()
    |> to_string()

  output =
    case day_of_year do
      1 ->
        "#{time} Happy New Year #{city}!"

      _ ->
        "time in #{city}: #{time}"
    end

  Mix.shell().info(output)
end
```

Now our failing test fails because the output doesn't match the new year greeting we got on the 31st of december. We can make this very testable by exctracting just that part out in a different function and testing that in isolation.

Let's create a new test for the new function we would like to see.

```elixir
test "the output contains a new years greeting on the first day of the year" do
  time = ~T[02:00:00.005]
  date = ~D[2019-01-01]
  {:ok, date_time} = NaiveDateTime.new(date, time)

  output = Mix.Tasks.Time.format_output(date_time, "Australia/Sydney")
  truncated_time = time |> Time.truncate(:second) |> to_string()

  assert output == truncated_time <> " Happy New Year Sydney!"
end
```

> I have hardcoded the timezone here for now, but that is good enough for now.

Reviewer: ```suggestion
I have hardcoded the timezone here, but that is good enough for now.
```

> We can make this test pass by creating the function extracting the output from the run function. We'll go ahead and extract the other conversions as private functions as well ending with the implementation below:

Reviewer: ```suggestion
We can make this test pass by extracting a few functions that handle formatting the output. We'll go ahead and extract the other conversions as private functions too, ending with the implementation below:
```

```elixir
def format_output(date_time, time_zone) do
  time = time(date_time)
  city = city(time_zone)

  case day_of_the_year(date_time) do
    1 ->
      "#{time} Happy New Year #{city}!"

    _ ->
      "time in #{city}: #{time}"
  end
end

defp city(time_zone) do
  String.split(time_zone, "/")
  |> Enum.drop(1)
  |> to_string()
end

defp day_of_the_year(date_time) do
  date_time
  |> DateTime.to_date()
  |> Date.day_of_year()
end

defp time(date_time) do
  date_time
  |> DateTime.to_naive
  |> NaiveDateTime.truncate(:second)
  |> NaiveDateTime.to_time()
  |> to_string()
end
```

> Now that was a lot more refactoring than I initially thought I would do, but we pulled this out into separate functions very nicely!

Reviewer: I agree this is a nice refactoring, but I'm not sure how it's relevant for the subject of this post?

Author: yes I agree. Maybe I can make the subject of this post New Years clean up? The OptionParser part is indeed the smallest part of the post.

What do you think?

Author: hmm or just name it refactoring, functional style and explain the pattern matching / arity matching of the functions?

As promised in the title of this post, let's talk about [OptionParser](https://hexdocs.pm/elixir/OptionParser.html). Basically `OptionParser` brings us some nice and convenient ways to add aliases and convert our arguments into options and switches.

As always let's test our way into our implementation. We want to be able to add a flag to our clock that will output the time in either 12- or 24-hour format. The current output uses 24 hour time, so we need to convert that to am / pm.

> Here is our test. I have created a helpser function to create the am or pm in our test.

Reviewer: ```suggestion
Here is our test. I have created a helper function to create the am or pm in our test.
```
